types/rdf-validate-shacl/index.d.ts

```diff
@@ -3,23 +3,25 @@
 // Definitions by: Tomasz Pluskiewicz <https://github.com/tpluscode>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 
-import * as RDF from 'rdf-js';
-import DataFactory = require('./src/data-factory');
+import { DataFactory, DatasetCore, DatasetCoreFactory } from 'rdf-js';
 import ValidationReport = require('./src/validation-report');
 
+type FactoryFor<D> = D extends DatasetCore<infer OutQuad, infer InQuad> ? DataFactory<OutQuad, InQuad> & DatasetCoreFactory<OutQuad, InQuad, D> : never;
+type OutQuadOf<T> = T extends DatasetCore<infer Q, any> ? Q : never;
+
 declare namespace SHACLValidator {
-    interface Options {
-        factory?: RDF.DataFactory & RDF.DatasetCoreFactory | undefined;
+    interface Options<F extends DataFactory & DatasetCoreFactory> {
+        factory?: F | undefined;
         maxErrors?: number | undefined;
     }
 }
 
-declare class SHACLValidator {
-    constructor(shapes: RDF.DatasetCore, options?: SHACLValidator.Options);
-    factory: DataFactory;
+declare class SHACLValidator<D extends DatasetCore, F extends FactoryFor<D>> {
+    constructor(shapes: D, options?: SHACLValidator.Options<F>);
+    factory: F;
     depth: number;
-    validate(data: RDF.DatasetCore): ValidationReport;
-    nodeConformsToShape(focusNode: RDF.Term, shapeNode: RDF.Term): boolean;
+    validate(data: D): ValidationReport<D>;
```

Comment:
Ditto, because `input dataset !== output dataset`

```suggestion
    validate(data: DatasetCore): ValidationReport<D>;
```